lib/Thruk/Utils/OAuth.pm

```diff
@@ -77,18 +77,42 @@ sub handle_oauth_login {
         } else {
             $ua->default_header(Authorization => "token ".$token->{"access_token"});
         }
-        _debug(sprintf("oauth login step2: fetching user id from: %s", $auth->{'api_url'})) if Thruk::Base->debug;
-        $res = $ua->get($auth->{'api_url'});
-        my $userinfo = _get_json($c, $res);
-        if(!$userinfo) {
-            return $c->detach_error({msg => "cannot fetch oauth user details", code => 500, debug_information => { res => $res }});
-        }
-        _debug(sprintf("oauth login step2: got user details:")) if Thruk::Base->debug;
-        _debug(Dumper($userinfo)) if Thruk::Base->debug;
-        # get username from response hash
-        my $login = $auth->{'login_field'} ? $userinfo->{$auth->{'login_field'}} : ($userinfo->{'login'} || $userinfo->{'email'});
-        if(!defined $login) {
-            return $c->detach_error({msg => "cannot find oauth user name", code => 500, debug_information => { userinfo => $userinfo }});
+        my $login;
+        if ($auth->{'api_url'}) {
+            _debug(sprintf("oauth login step2: fetching user id from: %s", $auth->{'api_url'})) if Thruk::Base->debug;
+            $res = $ua->get($auth->{'api_url'});
+            my $userinfo = _get_json($c, $res);
+            if(!$userinfo) {
+                return $c->detach_error({msg => "cannot fetch oauth user details", code => 500, debug_information => { res => $res }});
+            }
+            _debug(sprintf("oauth login step2: got user details:")) if Thruk::Base->debug;
+            _debug(Dumper($userinfo)) if Thruk::Base->debug;
+            # get username from response hash
+            $login = $auth->{'login_field'} ? $userinfo->{$auth->{'login_field'}} : ($userinfo->{'login'} || $userinfo->{'email'});
+            if(!defined $login) {
+                return $c->detach_error({msg => "cannot find oauth user name", code => 500, debug_information => { userinfo => $userinfo }});
+            }
+        } else {
+            # AD FS puts the claims in the id_token rather than in userinfo
+            _debug(sprintf("oauth login step2: api_url not set, looking for user id in id_token")) if Thruk::Base->debug;
+            use Crypt::JWT qw(decode_jwt);
```

Comment:
`use` will be evaluated during initial loading of the module, no matter where it is. You should use `require` if the module should be loaded during runtime.
Even better would be a require inside a eval block so you can return a useful error message if loading the module fails.